### PR TITLE
UCP/CORE: support cuda memory allocations with ucp_mem_map

### PR DESCRIPTION
**Pull request by Akshay-Venkatesh**
## What
Let ucp_mem_map allocate mem_type memory 

## Why ?
Needed for device staging buffers for rendezvous pipeline

### CODE REVIEW

**Comment by Akshay-Venkatesh** [comment]
@yosefe can you please review when possible?

File: src/ucp/core/ucp_mm.c [hunk 1]
```diff
@@ -192,6 +192,12 @@ static ucs_status_t ucp_mem_alloc(ucp_context_h context, size_t length,
     {
         method = context->config.alloc_methods[method_index].method;
 
+        /* Non-host memory allocation supported on METHOD_MD alone */
+        if ((memh->mem_type != UCS_MEMORY_TYPE_HOST)
+            && (method != UCT_ALLOC_METHOD_MD)) {
+            continue;
+        }
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:199]: what happens if we remove? it should be done already in UCT layer

File: src/ucp/core/ucp_mm.c [hunk 2]
```diff
@@ -415,6 +417,14 @@ ucs_status_t ucp_mem_map(ucp_context_h context, const ucp_mem_map_params_t *para
         goto out;
     }
 
+    if (((flags & UCP_MEM_MAP_FIXED) || (flags & UCP_MEM_MAP_NONBLOCK)) &&
+        ((params->field_mask & UCP_MEM_MAP_PARAM_FIELD_MEMORY_TYPE)
+         && (params->memory_type != UCS_MEMORY_TYPE_HOST))) {
+        ucs_error("UCP_MEM_MAP_FIXED/NONBLOCK unsupported with non-host mem");
+        status = UCS_ERR_INVALID_PARAM;
+        goto out;
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:425]: again, can we move this check to UCT layer?
also, NONBLOCK is just a hint; can ignore it for cuda mem

File: src/ucp/core/ucp_mm.c [hunk 3]
```diff
@@ -436,7 +446,8 @@ ucs_status_t ucp_mem_map(ucp_context_h context, const ucp_mem_map_params_t *para
     }
 
     if (flags & UCP_MEM_MAP_ALLOCATE) {
-        memory_type = UCS_MEMORY_TYPE_HOST;
+        memory_type = (params->field_mask & UCP_MEM_MAP_PARAM_FIELD_MEMORY_TYPE)
+                      ? params->memory_type : UCS_MEMORY_TYPE_HOST;
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.c:450]: UCP_PARAM_VALUE

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 4]
```diff
@@ -243,6 +243,43 @@ UCS_TEST_P(test_ucp_mmap, alloc) {
     }
 }
 
+UCS_TEST_P(test_ucp_mmap, alloc_cuda) {
+    std::vector<ucs_memory_type_t> mem_types = mem_buffer::supported_mem_types();
+    ucs_status_t status;
+    bool is_dummy;
+
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc:250]: remove one spaceline

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 5]
```diff
@@ -243,6 +243,43 @@ UCS_TEST_P(test_ucp_mmap, alloc) {
...
+
+    if (std::find(mem_types.begin(), mem_types.end(),
+                  UCS_MEMORY_TYPE_CUDA) == mem_types.end()) {
+        UCS_TEST_SKIP_R("cuda memory unsupported");
+    }
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc:255]: why not check all memory types?

**Review by yosefe — COMMENTED** [review]
pls check clang-format for code style issues

File: src/ucp/core/ucp_mm.c (hunk 1 above)
Comment (Akshay-Venkatesh) [reply src/ucp/core/ucp_mm.c]: removed

File: src/ucp/core/ucp_mm.c (hunk 2 above)
Comment (Akshay-Venkatesh) [reply src/ucp/core/ucp_mm.c]: removed

File: test/gtest/ucp/test_ucp_mmap.cc (hunk 5 above)
Comment (Akshay-Venkatesh) [reply test/gtest/ucp/test_ucp_mmap.cc]: added for all memtypes but it's possible that tests fail with ROCM/ROCM_MANAGED. I'm not sure if rocm has added alloc support in UCT yet.

File: test/gtest/ucp/test_ucp_mmap.cc [hunk 6]
```diff
@@ -243,6 +243,41 @@ UCS_TEST_P(test_ucp_mmap, alloc) {
     }
 }
 
+UCS_TEST_P(test_ucp_mmap, alloc_mem_type) {
+    ucs_status_t status;
+    bool is_dummy;
+    bool expect_rma_offload;
+
+    for (ucs_memory_type_t mem_type : mem_buffer::supported_mem_types()) {
+        for (int i = 0; i < 100 / ucs::test_time_multiplier(); ++i) {
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_mmap.cc:252]: minor: use ( )

**Review by yosefe — APPROVED** [review]

File: test/gtest/ucp/test_ucp_mmap.cc (hunk 6 above)
Comment (Akshay-Venkatesh) [reply test/gtest/ucp/test_ucp_mmap.cc]: Fixed

**Comment by Akshay-Venkatesh** [comment]
@yosefe Should I squash commits?

**Comment by yosefe** [comment]
> @yosefe Should I squash commits?

It seems there are some tests issues. Let's make sure all pass first.

**Comment by Akshay-Venkatesh** [comment]
> > @yosefe Should I squash commits?
> 
> It seems there are some tests issues. Let's make sure all pass first.

Ok will wait.

**Comment by Akshay-Venkatesh** [comment]
@petro-rudenko do you think the errors here are valid? 
https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/17760/logs/176
https://dev.azure.com/ucfconsort/0b36e3f0-8ab9-4a48-b68b-4b2350e02c88/_apis/build/builds/17760/logs/172

Seems to searching for transport 'abcd'...

**Comment by Akshay-Venkatesh** [comment]
@yosefe I'm unable to repro the following errors locally:
> [  FAILED  ] sysv/test_uct_event.am_no_fc/0, where GetParam() = sysv/memory
> [  FAILED  ] rc_mlx5/uct_p2p_rma_test_alloc_methods.xfer_reg_rcache/0, where GetParam() = rc_mlx5/mlx5_1:1

Would it be possible to rerun these tests?

**Comment by Akshay-Venkatesh** [comment]
@petro-rudenko Do you mind taking a look at this when you get a chance? They seem unrelated errors and may just need to be restarted.

https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=17904&view=logs&j=3af09b09-f681-502e-a77b-ab1dc5457b44&t=38d5da4a-ed53-5453-fe91-8988b34f7242&l=1682
https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=17904&view=logs&j=2d17f13d-f402-5448-dd96-eeac985c8863&t=61d25ca5-2380-522c-68e5-91463ebbd3bf&l=2228

**Comment by Akshay-Venkatesh** [comment]
@petro-rudenko There are still some errors on the Java side. Do you mind taking a look? https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=17904&view=logs&j=2d17f13d-f402-5448-dd96-eeac985c8863&t=61d25ca5-2380-522c-68e5-91463ebbd3bf&l=200

**Comment by petro-rudenko** [comment]
JUCX tests are passed now

**Comment by Akshay-Venkatesh** [comment]
> 
> 
> JUCX tests are passed now

Thanks @petro-rudenko. I realized that I was looking at older results as soon as I sent you the message.

**Comment by Akshay-Venkatesh** [comment]
@petro-rudenko 

Again, not sure if these are real issues:

https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=17935&view=logs&j=73bd2a9e-6688-5c22-d122-7b1a0b6cbb81&t=cd3d3953-51ff-5d5c-c5ca-2864f3c1d670&l=86608

Do you mind restarting cancelled tests?

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@Akshay-Venkatesh pls squash

**Pull request merged**